python/ql/lib/semmle/python/dataflow/new/internal/VariableCapture.qll

```diff
@@ -0,0 +1,207 @@
+/** Provides logic related to captured variables. */
+
+private import python
+private import DataFlowPublic
+private import semmle.python.dataflow.new.internal.DataFlowPrivate
+private import codeql.dataflow.VariableCapture as Shared
+
+// Note: The Javascript implementation (on the branch https://github.com/github/codeql/pull/14412)
+// had some tweaks related to performance. See these two commits:
+// - JS: Capture flow: https://github.com/github/codeql/pull/14412/commits/7bcf8b858babfea0a3e36ce61145954c249e13ac
+// - JS: Disallow consecutive captured contents: https://github.com/github/codeql/pull/14412/commits/46e4cdc6232604ea7f58138a336d5a222fad8567
+// The first is the main implementation, the second is a performance motivated restriction.
+// The restriction is to clear any `CapturedVariableContent` before writing a new one
+// to avoid long access paths (see the link for a nice explanation).
+private module CaptureInput implements Shared::InputSig<Location> {
+  private import python as PY
+
+  additional class ExprCfgNode extends ControlFlowNode {
+    ExprCfgNode() { isExpressionNode(this) }
+  }
+
+  class Callable extends Scope {
+    predicate isConstructor() { none() }
+  }
+
+  class BasicBlock extends PY::BasicBlock {
+    Callable getEnclosingCallable() { result = this.getScope() }
+
+    // Note `PY:BasicBlock` does not have a `getLocation`.
+    // (Instead it has a complicated location info logic.)
+    // Using the location of the first node is simple
+    // and we just need a way to identify the basic block
+    // during debugging, so this will be serviceable.
+    Location getLocation() { result = super.getNode(0).getLocation() }
+  }
+
+  BasicBlock getImmediateBasicBlockDominator(BasicBlock bb) { result = bb.getImmediateDominator() }
+
+  BasicBlock getABasicBlockSuccessor(BasicBlock bb) { result = bb.getASuccessor() }
+
+  class CapturedVariable extends LocalVariable {
+    Function f;
+
+    CapturedVariable() {
+      // note: captured variables originating on module scope is currently
+      // covered by global variable handling.
+      this.getScope() = f and
+      this.getAnAccess().getScope() != f
+    }
+
+    Callable getCallable() { result = f }
+
+    Location getLocation() { result = f.getLocation() }
+
+    /** Gets a scope that captures this variable. */
+    Scope getACapturingScope() {
+      result = this.getAnAccess().getScope().getScope*() and
+      result.getScope+() = f
+    }
+  }
+
+  class CapturedParameter extends CapturedVariable {
+    CapturedParameter() { this.isParameter() }
+
+    ControlFlowNode getCfgNode() { result.getNode().(Parameter) = this.getAnAccess() }
+  }
+
+  class Expr extends ExprCfgNode {
+    predicate hasCfgNode(BasicBlock bb, int i) { this = bb.getNode(i) }
+  }
+
+  class VariableWrite extends ControlFlowNode {
+    CapturedVariable v;
+
+    VariableWrite() { this = v.getAStore().getAFlowNode().(DefinitionNode).getValue() }
```

Comment:
I think this should really be the assignment node, as that is where the assignment takes place, and not the RHS of the assignment. Note that you will then have to change
```ql
result.(Flow::VariableWriteSourceNode).getVariableWrite() = n.(CfgNode).getNode()
```
in `asClosureNode` as well.

Reply:
As discussed offline, we will change this follow-up, once Python has gained cfg nodes for assignments.